File: Solutions/a60a_midp1_knapsack.cpp
/*
 * AUTHOR: Natchapol Srisang (UtopiaBeam)
 * KEYWORD: 0/1 Kanpsack
 */

#include <cstdio>
#include <vector>
#define N 500
using namespace std;

int dp[N+2][N+2];
int v[N+2], w[N+2];
vector<int> ans;

void ks(int n, int m) {
    if (n == 0 || m == 0) {
        return;
    }
    if (m >=w[n] && dp[n-1][m-w[n]] + v[n] == dp[n][m]) {
        ans.push_back(n);
        ks(n-1, m-w[n]);
    } else {
        ks(n-1, m);
    }
}

int main() {
    int n, m;
    scanf("%d %d", &n, &m);
    for (int i = 1; i <= n; i++) {
        scanf("%d", v+i);
    }
    for (int i = 1; i <= n; i++) {
        scanf("%d", w+i);
    }
    for (int i = 0; i <= n; i++) {
        for (int j = 0; j <= m; j++) {
            scanf("%d", &dp[i][j]);
        }
    }

    ks(n, m);

    printf("%d\n", ans.size());
    for (int &x: ans) {
        printf("%d ", x);
    }
    printf("\n");
    return 0;
}

File: Solutions/a60a_midp1_knapsack.py
#
# AUTHOR: Natchapol Srisang (UtopiaBeam)
# KEYWORD: 0/1 Kanpsack
#


def ks(n, m, ans=[]):
    if n == 0 or m == 0:
        print(len(ans))
        print(*ans)
        return
    global v, w, dp
    if m >= w[n-1] and dp[n-1][m-w[n-1]] + v[n-1] == dp[n][m]:
        ks(n-1, m-w[n-1], ans+[n])
    else:
        ks(n-1, m, ans)


n, m = map(int, input().split())
v, w = [list(map(int, input().split())) for _ in range(2)]
dp = [list(map(int, input().split())) for _ in range(n+1)]
ks(n, m)
